### PR TITLE
Factor out shared chain traversal when inbox and outbox handle new head.

### LINKED ISSUE #3052: [bug] Same actor and nonce but different cid
**Describe the bug**
This error has been seen in 0.2.4 and 0.3.0 version several times, when there are staged sectors waiting in line for sealing. 

`[0;37m16:29:28.595 [31mERROR [0;34m      node: [0mfailed to send commitSector message from t1kbn...yny to t2kab...xvi for sector with id 33....26: failed to add message to message pool: validation error adding message to pool: message pool contains message with same actor and nonce but different cid [0;37mnode.go:1110[0m`

And, I talked with some other miners, they got this too.

**To Reproduce**
Do not need to do any special steps, just send asks as fast as you can, and check the output log, then you will see this some time. 

**Expected behavior**
The nonce is increasing always, no same nonces for two messages in one miner's system.

**Version information**
Found in both test-devnet-0.2.2 and test-devnet-0.3.2

**Additional context**
Could there be race condition happened when multiple sectors for seal, especially when multiple workers are running?

### PR DESCRIPTION
**Pull request by anorth**
This is a step toward restoring lost messages to the outbox for #3052.

### CODE REVIEW

File: chain/traversal.go [hunk 1]
```diff
@@ -87,3 +87,33 @@ func TipSetProviderFromBlocks(ctx context.Context, blocks BlockProvider) TipSetP
 func (p *tipsetFromBlockProvider) GetTipSet(tsKey types.TipSetKey) (types.TipSet, error) {
 	return LoadTipSetBlocks(p.ctx, p.blocks, tsKey)
 }
+
+// CollectTipsToCommonAncestor traverses chains from two tipsets (called old and new) until their common
+// ancestor, collecting all tipsets that are in one chain but not the other.
+// The resulting lists of tipsets are ordered by decreasing height.
+func CollectTipsToCommonAncestor(ctx context.Context, store TipSetProvider, oldHead, newHead types.TipSet) (oldTips, newTips []types.TipSet, err error) {
```
Comment (frrist) [inline chain/traversal.go]: Mixed feelings the named returns :thinking:, I suppose this does improve readability.

File: core/outbox.go [hunk 2]
```diff
@@ -35,8 +36,8 @@ type Outbox struct {
 }
 
 type outboxChainProvider interface {
```
Comment (frrist) [inline core/outbox.go]: This is identical to the `inboxChainProvider` interface. Seeing as they both live in the same package, is there any reason to separate them like this?

File: core/policy.go [hunk 3]
```diff
@@ -81,14 +77,16 @@ func (p *DefaultQueuePolicy) HandleNewHead(ctx context.Context, target PolicyTar
...
-			log.Errorf("Outbound message %v expired un-mined after %d rounds", msg, p.maxAgeRounds)
+	if len(newTips) > 0 {
+		height, err := newTips[0].Height()
+		if err != nil {
+			return err
+		}
+		if height >= p.maxAgeRounds { // avoid uint subtraction overflow
+			expired := target.ExpireBefore(ctx, (height - p.maxAgeRounds))
+			for _, msg := range expired {
+				log.Errorf("Outbound message %v expired un-mined after %d rounds", msg, p.maxAgeRounds)
```
Comment (frrist) [inline core/policy.go]: (I know you haven't changed this here, but..) Does this warrant an error message -- could this be a warning instead?

File: go.sum [hunk 4]
```diff
@@ -1100,6 +1100,8 @@ golang.org/x/sys v0.0.0-20190712062909-fae7ac547cb7 h1:LepdCS8Gf/MVejFIt8lsiexZA
 golang.org/x/sys v0.0.0-20190712062909-fae7ac547cb7/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
 golang.org/x/sys v0.0.0-20190726091711-fc99dfbffb4e h1:D5TXcfTk7xF7hvieo4QErS3qqCB4teTffacDWr7CI+0=
 golang.org/x/sys v0.0.0-20190726091711-fc99dfbffb4e/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
+golang.org/x/sys v0.0.0-20190801041406-cbf593c0f2f3 h1:4y9KwBHBgBNwDbtu44R5o1fdOCQUEXhbk/P4A9WmJq0=
```
Comment (frrist) [inline go.sum]: I think these could be removed.

File: chain/traversal.go (hunk 1 above)
Comment (shannonwells) [reply chain/traversal.go]: Named returns are a good idea if you are returning multiples of the same type, that's my 2 centavos.

File: node/node.go [hunk 5]
```diff
@@ -732,20 +732,31 @@ func (node *Node) handleNewChainHeads(ctx context.Context, prevHead types.TipSet
 				log.Error("tipset of size 0 published on heaviest tipset channel. ignoring and waiting for a new heaviest tipset.")
 				continue
 			}
-
-			if err := node.Outbox.HandleNewHead(ctx, prevHead, newHead); err != nil {
-				log.Error("updating outbound message queue for new tipset", err)
+			if newHead.Equals(prevHead) {
+				log.Errorf("non-new head published on heaviest tipset channel, ignoring %s", newHead.Key())
+				continue
```
Comment (shannonwells) [inline node/node.go]: Similar question to above, should this be an error or warning?

**Pull request merged**